riscv/ctrl.py

```diff
@@ -0,0 +1,431 @@
+from myhdl import always_comb, modbv, concat, block
+
+from riscv.control_constants import *
+from riscv.csr_constants import *
+from riscv.alu_constants import *
+from riscv.opcode_constants import *
+from riscv.md_constants import *
+
+@block
+def vscale_ctrl(clk,
+                reset,
+                inst_DX,
+                imem_wait,
+                imem_badmem_e,
+                dmem_wait,
+                dmem_badmem_e,
+                cmp_true,
+                prv,
+                PC_src_sel,
+                imm_type,
+                bypass_rs1,
+                bypass_rs2,
+                src_a_sel,
+                src_b_sel,
+                alu_op,
+                dmem_en,
+                dmem_wen,
+                dmem_size,
+                dmem_type,
+                md_req_valid,
+                md_req_ready,
+                md_req_in_1_signed,
+                md_req_in_2_signed,
+                md_req_op,
+                md_req_out_sel,
+                md_resp_valid,
+                eret,
+                csr_cmd,
+                csr_imm_sel,
+                illegal_csr_access,
+                interrupt_pending,
+                interrupt_taken,
+                wr_reg_WB,
+                reg_to_wr_WB,
+                wb_src_sel_WB,
+                stall_IF,
+                kill_IF,
+                stall_DX,
+                kill_DX,
+                stall_WB,
+                kill_WB,
+                exception_WB,
+                exception_code_WB,
+                retire_WB):
+
+    # If Stage control
+    @always(clk.posedge)
+    def IF_stage():
+        if reset:
+            replay_IF = modbv(1)[1:]
+        else:
+            replay_IF = (redirect & imem_wait) | (fence_i & store_in_WB)
+
+    kill_IF = stall_IF | ex_IF | ex_DX | ex_WB | redirect | replay_IF | interrupt_taken
+    stall_IF = stall_DX | ((imem_wait & ~redirect) & ~(ex_WB | interrupt_taken))
+    ex_IF = imem_badmem_e & ~imem_wait & ~redirect & ~replay_IF
+
+    # DX Stage control
+
+    @always(clk.posedge)
+    def DX_stage():
+        if reset:
+            had_ex_DX = 0
+            prev_killed_DX = 0
+        elif ~stall_DX:
+            had_ex_DX = ex_IF
+            prev_killed_DX = kill_IF
+
+    # interrupts kill IF, DX instructions -- WB may commit
+    # Exceptions never show up falsely due to hazards -- don't get exceptions on stall
+
+    kill_DX   = stall_DX | ex_DX | x_WB | interrupt_taken
+    stall_DX  = stall_WB | ((load_use | raw_on_busy_md | (fence_i & store_in_WB) | (uses_md_unkilled & ~md_req_ready)) & ~(ex_DX | ex_WB | interrupt_taken))
+    new_ex_DX = ebreak | ecall | illegal_instruction | illegal_csr_access
+
+    killed_DX = prev_killed_DX | kill_DX
+    ex_DX     = had_ex_DX | new_ex_DX
+
+    @always_comb
+    def DX_stage_comb():
+        ex_code_DX = ECODE_INST_ADDR_MISALIGNED
+        if had_ex_DX:
+            ex_code_DX = ECODE_INST_ADDR_MISALIGNED
+        elif illegal_instruction:
```

Comment:
This doesn't seem right. You probably want to use .next instead of reassigning the variable. Does this module have a test?